Set Extruder Offsets on an LCD Controller and Save to EEPROM

gralco opened the pull request:
Same issue as thinkyhead mentioned on https://github.com/MarlinFirmware/Marlin/pull/2177

> I think we want to start labeling the extruders 1-4 instead of 0-3, at least in the user interface. Also I think we should try to minimize string storage. Instead of making 4 extruder strings, just append the number in code.

gralco committed:
Initial Extruder Offsets LCD and EEPROM

gralco committed:
Initial Extruder Offsets LCD and EEPROM

AnHardt commented on Marlin/language_en.h:
```diff
@@ -261,6 +261,24 @@
 #define MSG_ESTEPS                          "Esteps/mm"
 #endif
+#ifndef MSG_X1OFFSET
+#define MSG_X1OFFSET                         "T1X Offset"
```
Why 'T'? Why not 'E'?
We deal here with **E** xtruders not **T** emperatures.

Couldn't you generate E1X and only add "Offset" as a string?

gralco replied:
I used `T` because `M218` uses `T<extruder_number>`, I believe `T` stands for tool. I assume this is the case because Marlin can be used for standard CNCs as well.

thinkyhead replied:
That's an interesting point around the Marlin G-code, speaking of standards and practices. `T` is correct for the tool number. However, there's some question whether that's cool with G-code standards. Specifically, as I understand it, the G-code guidelines we're following reserve `G`, `M`, and `T` for the main codes, `N` for line numbers, `F` for feed-rate, `X`, `Y`, and `Z` for the three primary axes, etc.

That said, these "standards" are pretty loosely applied in Marlin. The `T` code is used by `M200` for the extruder number. `M204` and `M205` use it for "travel moves" rate. `M206` uses it for "theta." And a few other codes, such as `M221`, also use `T` for the extruder number.

But `M301` and `M303` use `E` for the extruder number. `G29` uses `E` as a flag for "enhanced G29."

So, for consistency, `T` seems best for the extruder number right now.

Wackerbarth replied:
> On May 28, 2015, at 8:23 PM, Scott Lahteine notifications@github.com wrote:
> That's an interesting point around the Marlin G-code, speaking of standards and practices. T is correct for the tool number. However, there's some question whether that's cool with G-code standards.
> 
> Which G-code standards? RS274?

3.7.3 Select Tool — T

To select a tool, program T..., where the T number is the carousel slot for the tool. The tool is not changed until an M6 is programmed (see Section 3.6.3).

3.6.3 Tool Change — M6

To change a tool in the spindle from the tool currently in the spindle to the tool most recently selected (using a T word — see Section 3.7.3), program M6.

> Specifically, as I understand it, the G-code guidelines we're following reserve G, M, and T for the main codes, N for line numbers, F for feed-rate, X, Y, and Z for the three primary axes, etc.
> 
> That said, these "standards" are pretty loosely applied in Marlin. The T code is used by M200 for the extruder number. M204 and M205 use it for "travel moves" rate. M206 uses it for "theta." And a few other codes, such as M221, use it for the extruder number.
> 
> But M301 and M303 use E for the extruder number. G29 uses E as a flag for "enhanced G29."
> 
> So, for consistency, T seems best for the extruder number right now.

thinkyhead replied:
> RS274?

Indeed. And which one? RS274 or RS264/NGC or RS274X or RS274D or RS274/VGER…

By "G-Code Standards" in this context, I mean the _de facto_ "standards" that RepRap 3D printer firmware has been following for a long time, apparently starting with Sprinter, but which is a very loose "g-code inspired" dialect and not apparently a real implementation of any well-defined standard.

AnHardt replied:
Interesting direction this discussion went. :-)

I wanted to talk about text read by humans in on the display, what has not necessary to be similar to a related g-code. And i wanted to talk about the efficiency of memory usage.

We already are in a sub-menu called "Extruder offset:". Should we now change to T.. for the menu-items? Maybe "T" is simply to short to be expressive. How about "Tool"?
If using "T" or "Tool" the sub-menu's title should be "Tool offset".

If it makes sense to split `"T1X Offset"` to `"T1X" " Offset"` or  `"T1X" MSG_OFFSET` entirely depends on the question when(where) the concatenation is done. I suppose all versions will end as a PSTR("T1X Offset") because the concatenation is done by the preprocessor and no memory is saved. But here is not the right place to discuss this.

For short. I suggest to ether rename the the menu-title or the menu-items.

The 'T' as a parameter-name for the g-code is ok for me, since the hosts/slices are used to this usage now. (If a eventually coming "standard" will forbid this usage for 'T' we can change this - but i don't believe on that.)

gralco replied:
I think we can say with confidence that the current quasi-standard has been defined by various firmware projects as well as Skeinforge.

When I picked `M19` and `G5` I simply picked them on the basis of their availability and the neighboring numbers, luckily these functions are not integral for slicers.

But the real question is, who should drive the standards and who should conform, the firmwares or the slicers? Maybe some sort of consortium should be formed that composes these standards, something like the Open Handset Alliance (there's currently no W3C equivalent for reprap).

thinkyhead replied:
> defined by various firmware projects as well as Skeinforge

Oh yes, we should not forget their immense contribution, bridging the worlds of CNC and RepRap.

thinkyhead replied:
> When I picked M19 and G5…

I think that's a smart approach because related codes should live together, but lately I'm also checking the RepRap wiki G-code page to see if there's either some code that does what I want, but hasn't been implemented yet, or to see if the code I would like to use already has some well-defined usage.

thinkyhead commented:
> Interesting direction this discussion went.

There's got to be one of those "engineering laws" that states: Every engineering discussion eventually leads to the topic of Standards.

boelle commented:
the build failed, please rebase on latest devlopment

gralco committed:
Add ifdef EXTRUDER_OFFSET_X/Y here

gralco committed:
Add ifdef EXTRUDER_OFFSET_X/Y here

gralco committed:
tmp4 and tmp5 still need to be defined here even if EXTRUDER_OFFSET_X/Y are not defined.

gralco committed:
tmp4 and tmp5 still need to be defined here even if EXTRUDER_OFFSET_X/Y are not defined.

gralco committed:
fix indent

gralco committed:
fix indent

gralco committed:
The arrays have two elements each.

gralco committed:
The arrays have two elements each.

gralco commented:
Rebased and resolved conflicts.

mkile committed:
Slightly updated translation

Slightly updated translation.

mkile committed:
Slightly updated translation

Slightly updated translation.

mkile committed:
Update language_ru.h

mkile committed:
Update language_ru.h

gralco committed:
Resolving more conflicts.

gralco committed:
Resolving more conflicts.

gralco committed:
Forgot to remove "<<<<<<< HEAD".

thinkyhead committed:
Rename SDCARDDETECT to SDCARDDETECT_PIN

- Rename the pin so it can be tested with `PIN_EXISTS`
- Fix some incorrect tests for `SDCARDDETECT`

thinkyhead committed:
Tweaks to pins comments

thinkyhead committed:
SDCARDDETECT -> SD_DETECT, added docs

- `SD_DETECT_PIN` replaces `SDCARDDETECT`
- `SD_DETECT_INVERTED` replaces `SDCARDDETECTINVERTED`
- Revise the description of `SD_DETECT_INVERTED`
- Add a note about the override of `SD_DETECT_INVERTED` in
`Conditionals.h`

thinkyhead committed:
In for a pound, spacing adjustments

thinkyhead committed:
Add sanity checks for old names

thinkyhead committed:
Group the fail case in one negation

thinkyhead committed:
Merge pull request #2572 from thinkyhead/autostart_tweak

Tweak logic in checkautostart

Wackerbarth committed:
Development Branch DEPRECATED

thinkyhead committed:
Merge pull request #2569 from thinkyhead/rename_sdcarddetect

Rename SDCARDDETECT to SD_DETECT_PIN

thinkyhead committed:
Merge pull request #2579 from djabi/Development

Development

gralco committed:
Fix conflicts

gralco committed:
Merge branch 'Extruder-Offsets-LCD-and-EEPROM' of https://github.com/gralco/Marlin into Extruder-Offsets-LCD-and-EEPROM

gralco committed:
Fix conflicts
